docs/proposals/apb-state-support.md

```diff
@@ -0,0 +1,60 @@
+# Abstract
+
+This proposal will outline a simple and secure method for allowing APB developers to set state during a APB
+action and then be able to access that state during a subsequent APB action.
+
+## Actions
+   - provision
+   - deprovision
+   - update
+   - bind 
+   - unbind
+   
+   
+## Problem Description
+
+APBs are stateless. All state is managed for them by the broker. Currently the broker passes in parameters specified
+during a request to the catalog. It also passes in some additional parameters such as the namespace etc, additionally 
+we also pass in credentials created during the provision. While this is useful, there is no mechanism for an APB to store 
+and access data across actions without exposing it to the end user. We want to avoid APB developers working around this 
+limitation by doing suboptimal things such as storing extra data in the credentials or by creating a ConfigMap in the users 
+namespace as both of these expose the data to the user and as the namespace is controlled by the user, it is naturally 
+not trustworthy or reliable.
+
+
+
+# Proposed Solution
+
+## APB module for handling state
+
+This module will expose an API to the APB developer for setting and getting key value pairs.
+
+```
+
+- name: Save some stuff
+  asb_set_state:
+    service_name: "{{ service_name }}"
+
+```
+
+Under the hood this APB module would take the key value pair, and store it in a
+ConfigMap (or possibly a secret) labelled ```apb:state```. This ConfigMap would live within the 
+temporary namespace where the APB was running. 
+
+## Update broker to manage APB created state ConfigMaps
+
+To ensure the state was persisted across APB action, the broker would watch for ConfigMaps being added in the 
+APB namespace with the expected label. When a new ConfigMap was created in the APB's namespace, the broker would 
+copy this ConfigMap over to the broker's namepsace and name it ```<ServiceInstanceID>-state```. If a ConfigMap with 
+that name was already present, the broker would update and append the values. The broker would also watch 
+for ``Modified`` events on the ConfigMap and ensure to update the corresponding ConfigMap in the broker's namespace. 
+There should only ever be one ConfigMap per ServiceInstance. The ConfigMap would be removed from the broker's namespace 
+once the ServiceInstance was deleted. 
+
+## Update broker to pass through initial state to APB
+
+For every APB action, except provision as there would be no state at this point, if a ConfigMap (ServiceInstanceID-state) is present, 
+in the broker's namespace, its key value pairs will be passed through to the APB prefixed with ```state_<key> = value ```
```

Comment:
I wonder if this should actually be that the configmap/secret gets copied into the transient namespace and we add it to the pod. The contract then just has to define where the state gets mounted and it is up to the bundle to use that?

Reply:
Originally, I thought that this Proposal was suggesting that the APB (or service bundle) has the option to create a configmap that the broker would save for subsequent runs for that service instance. However, if it were possible for the broker to create a configmap on behalf of the APB (or service bundle) mounted at `/bundle/state` that the bundle could use for saving state, that sounds even better to me (if it is possible).

Reply:
@maleck13 I'm curious your thoughts on the above. If the broker were to provide Service Bundles a mount point `/bundle/state` as an example for saving state information, is that keeping with what you had in mind?

I'm actually curious if the broker could create a configmap in the `ansible-service-broker` namespace and provide that as a volume to an APB pod in the sandbox namespace.

Reply:
I don't think it is possible to mount a configmap from one namespace into a pod in another namespace . 
However it would be possible for the broker to create a configmap in the transient namespace during a provision and for the service bundle to be able to save state to it, then right before clean up the broker would copy this configmap to the broker's namespace. During any follow on action for that serviceinstance, the broker could then create a copy of the configmap in the transient namespace and add a volume to the pod. The service bundle could then add more state if needed and access any previous state from the mount point. 

How would you see the service bundle accessing the state? would this be via a ```apb_get_state <key>``` or just directly accessing the mount location?

Reply:
> mounted at /bundle/state that the bundle could use for saving state, that sounds even better to me (if it is possible).
Not sure this is possible. I don't think the contents of the configmap would be updated just by saving the state here.

Reply:
I played around with configmaps in pods a little and see that it is a read-only file system. I should have known better about mapping configmaps across namespaces...I've done worse :sunglasses:.

Reply:
I think we may be back to @shawn-hurley 's original comment. Instead of passing through the key/value pairs could we:

1. Create the configmap `$POD_NAME` in `$POD_NAMESPACE` with the current set of key/value pairs
1. Mount the configmap, read-only :sunglasses:, to a known location on the image for reading those values